### ISSUE TITLE
pipe fails to continue with no matched route

### ISSUE THREAD

**Issue by fabiocarneiro** [zendframework/zend-expressive#146]
The problem occurs when I try to return add a JsonResponse to the $next callable in pre_routing. Having something in post_routing doesn't really make any difference but I was expecting the request to reach a middleware there.

https://gist.github.com/fabiocarneiro/81ea4e1eb0a52b8ded12

```
2015/10/05 15:32:17 [error] 15#0: *320 FastCGI sent in stderr: "PHP message: PHP Catchable fatal error:  Argument 4 passed to Zend\Stratigility\Dispatch::__invoke() must be an instance of Zend\Stratigility\Http\Response, instance of Zend\Diactoros\Response\JsonResponse given, called in /usr/share/nginx/html/vendor/zendframework/zend-stratigility/src/Next.php on line 113 and defined in /usr/share/nginx/html/vendor/zendframework/zend-stratigility/src/Dispatch.php on line 53" while reading response header from upstream, client: 172.17.42.1, server: , request: "POST /login HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: "localhost", referrer: "http://localhost:63342/frontend/index.html"
```

**Comment by fabiocarneiro** [comment]
@weierophinney 
When I return Response from route middlewares, it doesn't reach the middlewares in post routing. I don't know if that is a proper way of handling that, but I'd suggest to have at least post_routing as separate pipe, that you reach even if you return response in routing middlewares (otherwise you almost never is able to return Response, but always need to use the $next, expecting that if in the future you want to add post_middlewares, you'll be able to reach them).

If I ```return $next($request, new \Zend\Diactoros\Response\JsonResponse());``` it never works. It only works with ```$next($request, new \Zend\Diactoros\Stratigility\Response(new \Zend\Diactoros\Response\JsonResponse()));```

**Comment by danizord** [comment]
I think https://github.com/zendframework/zend-stratigility/pull/32 fixes this one

**Comment by weierophinney** [comment]
@fabiocarneiro Can you run:

```bash
$ composer update zendframework/zend-stratigility
```

and test? The fix in zendframework/zend-stratigility#32 should resolve the reported issue.

In terms of always executing post-pipeline middleware, that's not the way Stratigility works. If you want to accomplish "post" middleware that runs *always*, the way to do that is to wrap your Expressive application in other middleware:

```php
function ($req, $res, $next) use ($app) {
    $response = $app->run($req, $res);
    // do what you want here
    return $response;
}
```

That middleware could potentially be another Expressive application, if desired.
